aesara/tensor/basic.py

```diff
@@ -1302,17 +1302,26 @@ def eye(n, m=None, k=0, dtype=None):
 
     Returns
     -------
-    ndarray of shape (N,M)
-        An array where all elements are equal to zero, except for the `k`-th
-        diagonal, whose values are equal to one.
-
+    aesara tensor of shape (N,M)
+        A symbolic tensor representing a matrix where all elements are equal to zero,
+        except for the `k`-th diagonal, whose values are equal to one.
     """
-    if dtype is None:
-        dtype = config.floatX
+
     if m is None:
         m = n
-    localop = Eye(dtype)
-    return localop(n, m, k)
+    if dtype is None:
+        dtype = aesara.config.floatX
+
+    n = as_tensor_variable(n)
+    m = as_tensor_variable(m)
+    k = as_tensor_variable(k)
+
+    i = switch(k >= 0, k, -k * m)
```

Comment:
We should consider using `ifelse` here&mdash;assuming that it's at least a little more efficient and/or suitable for this scalar-only case.

Reply:
`ifelse` seems slightly less performant, here are my tests:

```python
def eye_control_test(n, m, k, i_func):
    n = at.as_tensor_variable(n, name='n')
    m = at.as_tensor_variable(m, name='m')
    k = at.as_tensor_variable(k, name='k')

    i = i_func(k >= 0, k, -k * m)

    eye = at.zeros(n * m, dtype='float64')
    eye = at.subtensor.set_subtensor(eye[i:(m - k)*m:m + 1], 1).reshape((n, m))
    eye = at.subtensor.set_subtensor(eye[m - k:, :], 0)
    
    return eye

eye_sw = aesara.function([n_at,m_at,k_at], [eye_control_test(n_at, m_at, k_at, at.switch)], profile=True)
eye_ife = aesara.function([n_at,m_at,k_at], [eye_control_test(n_at, m_at, k_at, aesara.ifelse.ifelse)], profile=True)
```

```
%timeit eye_sw(n, m, k)
71.4 µs ± 300 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
````

```
%timeit eye_ife(n, m, k)
80.3 µs ± 365 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

<details><summary>Switch profile output </summary>

<p>

```
Function profiling
==================
  Message: C:\Users\Jesse\AppData\Local\Temp\ipykernel_3880\3286388574.py:1
  Time in 81111 calls to Function.__call__: 4.435331e+00s
  Time in Function.vm.__call__: 0.5880916118621826s (13.259%)
  Time in thunks: 0.3369591236114502s (7.597%)
  Total compilation time: 1.466007e-01s
    Number of Apply nodes: 14
    Aesara rewrite time: 2.693057e-02s
       Aesara validate time: 0.000000e+00s
    Aesara Linker time (includes C, CUDA code generation/compiling): 0.0049839019775390625s
       Import time 0.000000e+00s
       Node make_thunk time 4.983902e-03s
           Node IncSubtensor{InplaceSet;int32::, ::}(Reshape{2}.0, TensorConstant{0}, ScalarFromTensor.0) time 1.021385e-03s
           Node Elemwise{mul,no_inplace}(n, m) time 9.975433e-04s
           Node ScalarFromTensor(Elemwise{add,no_inplace}.0) time 9.973049e-04s
           Node Elemwise{Composite{Switch(GE(i0, i1), i0, (i2 * i0 * i3))}}(k, TensorConstant{0}, TensorConstant{-1}, m) time 9.968281e-04s
           Node ScalarFromTensor(Elemwise{Mul}[(0, 0)].0) time 9.708405e-04s

Time in all call to aesara.grad() 0.000000e+00s
Time since aesara import 17110.281s
Class
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Class name>
  30.2%    30.2%       0.102s       6.27e-07s     C   162222       2   aesara.tensor.subtensor.IncSubtensor
  26.3%    56.5%       0.089s       1.09e-06s     C    81111       1   aesara.tensor.basic.Alloc
  18.4%    74.9%       0.062s       1.52e-07s     C   405555       5   aesara.tensor.elemwise.Elemwise
  16.9%    91.7%       0.057s       7.01e-07s     C    81111       1   aesara.tensor.shape.Reshape
   5.9%    97.6%       0.020s       6.14e-08s     C   324444       4   aesara.tensor.basic.ScalarFromTensor
   2.4%   100.0%       0.008s       9.84e-08s     C    81111       1   aesara.tensor.basic.MakeVector
   ... (remaining 0 Classes account for   0.00%(0.00s) of the runtime)

Ops
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Op name>
  26.3%    26.3%       0.089s       1.09e-06s     C     81111        1   Alloc
  18.9%    45.3%       0.064s       7.86e-07s     C     81111        1   IncSubtensor{InplaceSet;int32:int32:int32}
  16.9%    62.1%       0.057s       7.01e-07s     C     81111        1   Reshape{2}
  11.2%    73.4%       0.038s       4.67e-07s     C     81111        1   IncSubtensor{InplaceSet;int32::, ::}
   5.9%    79.3%       0.020s       6.14e-08s     C     324444        4   ScalarFromTensor
   5.6%    84.9%       0.019s       2.34e-07s     C     81111        1   Elemwise{mul,no_inplace}
   4.1%    89.1%       0.014s       1.72e-07s     C     81111        1   Elemwise{Composite{Switch(GE(i0, i1), i0, (i2 * i0 * i3))}}
   3.5%    92.6%       0.012s       1.47e-07s     C     81111        1   Elemwise{add,no_inplace}
   3.3%    95.9%       0.011s       1.35e-07s     C     81111        1   Elemwise{sub,no_inplace}
   2.4%    98.2%       0.008s       9.84e-08s     C     81111        1   MakeVector{dtype='int32'}
   1.8%   100.0%       0.006s       7.38e-08s     C     81111        1   Elemwise{Mul}[(0, 0)]
   ... (remaining 0 Ops account for   0.00%(0.00s) of the runtime)

Apply
------
<% time> <sum %> <apply time> <time per call> <#call> <id> <Apply name>
  26.3%    26.3%       0.089s       1.09e-06s   81111     5   Alloc(TensorConstant{0.0}, Elemwise{mul,no_inplace}.0)
  18.9%    45.3%       0.064s       7.86e-07s   81111    11   IncSubtensor{InplaceSet;int32:int32:int32}(Alloc.0, TensorConstant{1}, ScalarFromTensor.0, ScalarFromTensor.0, ScalarFromTensor.0)
  16.9%    62.1%       0.057s       7.01e-07s   81111    12   Reshape{2}(IncSubtensor{InplaceSet;int32:int32:int32}.0, MakeVector{dtype='int32'}.0)
  11.2%    73.4%       0.038s       4.67e-07s   81111    13   IncSubtensor{InplaceSet;int32::, ::}(Reshape{2}.0, TensorConstant{0}, ScalarFromTensor.0)
   5.6%    79.0%       0.019s       2.34e-07s   81111     1   Elemwise{mul,no_inplace}(n, m)
   4.1%    83.1%       0.014s       1.72e-07s   81111     4   Elemwise{Composite{Switch(GE(i0, i1), i0, (i2 * i0 * i3))}}(k, TensorConstant{0}, TensorConstant{-1}, m)
   3.5%    86.7%       0.012s       1.47e-07s   81111     2   Elemwise{add,no_inplace}(TensorConstant{1}, m)
   3.3%    89.9%       0.011s       1.35e-07s   81111     3   Elemwise{sub,no_inplace}(m, k)
   2.4%    92.3%       0.008s       9.84e-08s   81111     0   MakeVector{dtype='int32'}(n, m)
   2.4%    94.7%       0.008s       9.84e-08s   81111     8   ScalarFromTensor(Elemwise{Composite{Switch(GE(i0, i1), i0, (i2 * i0 * i3))}}.0)
   1.8%    96.5%       0.006s       7.38e-08s   81111     9   Elemwise{Mul}[(0, 0)](Elemwise{sub,no_inplace}.0, m)
   1.5%    97.9%       0.005s       6.15e-08s   81111     7   ScalarFromTensor(Elemwise{sub,no_inplace}.0)
   1.2%    99.1%       0.004s       4.90e-08s   81111    10   ScalarFromTensor(Elemwise{Mul}[(0, 0)].0)
   0.9%   100.0%       0.003s       3.69e-08s   81111     6   ScalarFromTensor(Elemwise{add,no_inplace}.0)
   ... (remaining 0 Apply instances account for 0.00%(0.00s) of the runtime)

Here are tips to potentially make your code run faster
                 (if you think of new ones, suggest them on the mailing list).
                 Test them first, as they are not guaranteed to always provide a speedup.
  - Try the Aesara flag floatX=float32
```

</p>

</details>


<details><summary> `ifelse` profile output</summary>

<p>

```
Function profiling
==================
  Message: C:\Users\Jesse\AppData\Local\Temp\ipykernel_3880\3286388574.py:2
  Time in 81111 calls to Function.__call__: 5.012699e+00s
  Time in Function.vm.__call__: 1.0410208702087402s (20.768%)
  Time in thunks: 0.7277634143829346s (14.518%)
  Total compilation time: 1.456079e-01s
    Number of Apply nodes: 16
    Aesara rewrite time: 2.695537e-02s
       Aesara validate time: 0.000000e+00s
    Aesara Linker time (includes C, CUDA code generation/compiling): 0.005983829498291016s
       Import time 0.000000e+00s
       Node make_thunk time 5.983829e-03s
           Node Reshape{2}(IncSubtensor{InplaceSet;int32:int32:int32}.0, MakeVector{dtype='int32'}.0) time 1.024008e-03s
           Node Elemwise{Mul}[(0, 0)](Elemwise{sub,no_inplace}.0, m) time 9.980202e-04s
           Node Elemwise{mul,no_inplace}(n, m) time 9.975433e-04s
           Node IncSubtensor{InplaceSet;int32:int32:int32}(Alloc.0, TensorConstant{1}, ScalarFromTensor.0, ScalarFromTensor.0, ScalarFromTensor.0) time 9.970665e-04s
           Node Elemwise{mul,no_inplace}(TensorConstant{-1}, k, m) time 9.965897e-04s

Time in all call to aesara.grad() 0.000000e+00s
Time since aesara import 16971.512s
Class
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Class name>
  55.6%    55.6%       0.405s       2.49e-06s     Py  162222       1   aesara.ifelse.IfElse
  13.3%    68.9%       0.097s       5.96e-07s     C   162222       2   aesara.tensor.subtensor.IncSubtensor
  10.6%    79.5%       0.077s       9.48e-07s     C    81111       1   aesara.tensor.basic.Alloc
   9.9%    89.3%       0.072s       1.77e-07s     C   405555       6   aesara.tensor.elemwise.Elemwise
   6.2%    95.5%       0.045s       5.54e-07s     C    81111       1   aesara.tensor.shape.Reshape
   3.0%    98.5%       0.022s       6.75e-08s     C   324444       4   aesara.tensor.basic.ScalarFromTensor
   1.5%   100.0%       0.011s       1.35e-07s     C    81111       1   aesara.tensor.basic.MakeVector
   ... (remaining 0 Classes account for   0.00%(0.00s) of the runtime)

Ops
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Op name>
  55.6%    55.6%       0.405s       2.49e-06s     Py    162222        1   if{inplace}
  10.6%    66.2%       0.077s       9.48e-07s     C     81111        1   Alloc
   9.2%    75.4%       0.067s       8.24e-07s     C     81111        1   IncSubtensor{InplaceSet;int32:int32:int32}
   6.2%    81.5%       0.045s       5.54e-07s     C     81111        1   Reshape{2}
   4.1%    85.6%       0.030s       3.68e-07s     C     81111        1   IncSubtensor{InplaceSet;int32::, ::}
   3.0%    88.6%       0.022s       6.75e-08s     C     324444        4   ScalarFromTensor
   2.9%    91.5%       0.021s       2.58e-07s     C     81111        2   Elemwise{mul,no_inplace}
   2.6%    94.1%       0.019s       2.33e-07s     C     81111        1   Elemwise{ge,no_inplace}
   1.9%    96.0%       0.014s       1.72e-07s     C     81111        1   Elemwise{sub,no_inplace}
   1.5%    97.5%       0.011s       1.35e-07s     C     81111        1   MakeVector{dtype='int32'}
   1.4%    98.9%       0.010s       1.23e-07s     C     81111        1   Elemwise{Mul}[(0, 0)]
   1.1%   100.0%       0.008s       9.84e-08s     C     81111        1   Elemwise{add,no_inplace}
   ... (remaining 0 Ops account for   0.00%(0.00s) of the runtime)

Apply
------
<% time> <sum %> <apply time> <time per call> <#call> <id> <Apply name>
  55.6%    55.6%       0.405s       2.49e-06s   162222     9   if{inplace}(Elemwise{ge,no_inplace}.0, k, Elemwise{mul,no_inplace}.0)
  10.6%    66.2%       0.077s       9.48e-07s   81111     6   Alloc(TensorConstant{0.0}, Elemwise{mul,no_inplace}.0)
   9.2%    75.4%       0.067s       8.24e-07s   81111    13   IncSubtensor{InplaceSet;int32:int32:int32}(Alloc.0, TensorConstant{1}, ScalarFromTensor.0, ScalarFromTensor.0, ScalarFromTensor.0)
   6.2%    81.5%       0.045s       5.54e-07s   81111    14   Reshape{2}(IncSubtensor{InplaceSet;int32:int32:int32}.0, MakeVector{dtype='int32'}.0)
   4.1%    85.6%       0.030s       3.68e-07s   81111    15   IncSubtensor{InplaceSet;int32::, ::}(Reshape{2}.0, TensorConstant{0}, ScalarFromTensor.0)
   2.9%    88.5%       0.021s       2.58e-07s   81111     1   Elemwise{mul,no_inplace}(n, m)
   2.6%    91.1%       0.019s       2.33e-07s   81111     5   Elemwise{ge,no_inplace}(k, TensorConstant{0})
   1.9%    93.0%       0.014s       1.72e-07s   81111     3   Elemwise{sub,no_inplace}(m, k)
   1.5%    94.5%       0.011s       1.35e-07s   81111     0   MakeVector{dtype='int32'}(n, m)
   1.4%    95.9%       0.010s       1.23e-07s   81111    10   Elemwise{Mul}[(0, 0)](Elemwise{sub,no_inplace}.0, m)
   1.2%    97.1%       0.009s       1.10e-07s   81111    11   ScalarFromTensor(if{inplace}.0)
   1.1%    98.2%       0.008s       9.84e-08s   81111     2   Elemwise{add,no_inplace}(TensorConstant{1}, m)
   1.0%    99.2%       0.007s       8.62e-08s   81111     8   ScalarFromTensor(Elemwise{sub,no_inplace}.0)
   0.5%    99.7%       0.004s       4.92e-08s   81111     7   ScalarFromTensor(Elemwise{add,no_inplace}.0)
   0.3%   100.0%       0.002s       2.46e-08s   81111    12   ScalarFromTensor(Elemwise{Mul}[(0, 0)].0)
   ... (remaining 0 Apply instances account for 0.00%(0.00s) of the runtime)

Here are tips to potentially make your code run faster
                 (if you think of new ones, suggest them on the mailing list).
                 Test them first, as they are not guaranteed to always provide a speedup.
  - Try the Aesara flag floatX=float32
```

</p>
</details>

I don't see any control flow operations in the `switch` profile output, is it nested (as with `scan`)?

Reply:
I don't know if it's what you're referring to, but while doing more testing I discovered the 2nd `set_subtensor` isn't robust to certain values of `k > m`. The following makes it correct in all cases:

```python
    i = at.switch(at.ge(k, 0), k, -k * m)
    last_row = at.maximum(0, m-k)
    
    eye = at.zeros(n * m, dtype='float64')
    eye = at.subtensor.set_subtensor(eye[i::m + 1], 1).reshape((n, m))
    eye = at.subtensor.set_subtensor(eye[last_row:, :], 0)
```

Reply:
https://github.com/aesara-devs/aesara/pull/1224

Reply:
Here's the graph for the `Switch` variant of `eye` in your example:
```python
IncSubtensor{InplaceSet;int32::, ::} [id A] 13
 |Reshape{2} [id B] 12
 | |IncSubtensor{InplaceSet;int32:int32:int32} [id C] 11
 | | |Alloc [id D] 5
 | | | |TensorConstant{0.0} [id E]
 | | | |Elemwise{mul,no_inplace} [id F] 1
 | | |   |n [id G]
 | | |   |m [id H]
 | | |TensorConstant{1} [id I]
 | | |ScalarFromTensor [id J] 8
 | | | |Elemwise{Composite{Switch(GE(i0, i1), i0, (i2 * i0 * i3))}} [id K] 4
 | | |   |k [id L]
 | | |   |TensorConstant{0} [id M]
 | | |   |TensorConstant{-1} [id N]
 | | |   |m [id H]
 | | |ScalarFromTensor [id O] 10
 | | | |Elemwise{Mul}[(0, 0)] [id P] 9
 | | |   |Elemwise{sub,no_inplace} [id Q] 3
 | | |   | |m [id H]
 | | |   | |k [id L]
 | | |   |m [id H]
 | | |ScalarFromTensor [id R] 6
 | |   |Elemwise{add,no_inplace} [id S] 2
 | |     |TensorConstant{1} [id T]
 | |     |m [id H]
 | |MakeVector{dtype='int32'} [id U] 0
 |   |n [id G]
 |   |m [id H]
 |TensorConstant{0} [id M]
 |ScalarFromTensor [id V] 7
   |Elemwise{sub,no_inplace} [id Q] 3
```
and the `IfElse` variant:
```python
IncSubtensor{InplaceSet;int32::, ::} [id A] 15
 |Reshape{2} [id B] 14
 | |IncSubtensor{InplaceSet;int32:int32:int32} [id C] 13
 | | |Alloc [id D] 6
 | | | |TensorConstant{0.0} [id E]
 | | | |Elemwise{mul,no_inplace} [id F] 1
 | | |   |n [id G]
 | | |   |m [id H]
 | | |TensorConstant{1} [id I]
 | | |ScalarFromTensor [id J] 11
 | | | |if{inplace} [id K] 9
 | | |   |Elemwise{ge,no_inplace} [id L] 5
 | | |   | |k [id M]
 | | |   | |TensorConstant{0} [id N]
 | | |   |k [id M]
 | | |   |Elemwise{mul,no_inplace} [id O] 4
 | | |     |TensorConstant{-1} [id P]
 | | |     |k [id M]
 | | |     |m [id H]
 | | |ScalarFromTensor [id Q] 12
 | | | |Elemwise{Mul}[(0, 0)] [id R] 10
 | | |   |Elemwise{sub,no_inplace} [id S] 3
 | | |   | |m [id H]
 | | |   | |k [id M]
 | | |   |m [id H]
 | | |ScalarFromTensor [id T] 7
 | |   |Elemwise{add,no_inplace} [id U] 2
 | |     |TensorConstant{1} [id V]
 | |     |m [id H]
 | |MakeVector{dtype='int32'} [id W] 0
 |   |n [id G]
 |   |m [id H]
 |TensorConstant{0} [id N]
 |ScalarFromTensor [id X] 8
   |Elemwise{sub,no_inplace} [id S] 3
```

There are a couple things to consider in this output:
1. `n`, `m`, and `k` are scalars, but they're represented as tensors (i.e. through the use of `Elemwise` and `ScalarFromTensor`).  I've never been comfortable with the use of this unnecessary array machinery in cases like this (e.g. calculations that only ever involve scalars), but I've also never measured the performance differences, so I don't really know if there is a problem here.  If there is, we might be able to add a rewrite that removes the `Elemwise`s, especially in these cases.
2. The `k >= 0` and `-k * m` are "fused" with the `Switch` into a single scalar C implementation that likely has a lot less "context switching"/call overhead than the `IfElse` case.  This combined with 1., under the assumption that there is measurable `Elemwise` overhead compared to the scalar-only case, could amount to something significant.

Reply:
Regardless, using `ifelse` doesn't seem like an improvement at the moment, so this question is resolved.